### DISCUSSION TITLE
Ability to assign to a file extension a way to process that type of a file

### DISCUSSION THREAD

**Discussion by kamoshi** [withastro/roadmap#813] [Proposal]
# Summary

> [!NOTE]  
> This proposal is effectively a superset of https://github.com/withastro/roadmap/discussions/503

There should be a way to set via astro config, what types of files should be processed in what way. For example `.md` by MarkDoc processor, or `.lhs` by regular Markdown processor, or `.adoc` by a custom AsciiDoc processor.

# Background & Motivation

It is quite common to think of a file extension as a concrete category of file, but sadly this becomes much more muddy as you start to consider the broader ecosystem and interop between different software.

For example, some users have previously brought up the problem of not being able to use `.mdoc` files in some software (Obsidian, etc.), and there being no way to tell Astro to treat `.md` files as MarkDoc. Another example are `.lhs` (literate Haskell) files, which allow you to write Haskell source code formatted like LaTeX or Markdown. Literate Haskell can be a Markdown document, however you are not able to use this type of file in Astro, because there is currently no way to tell Astro to treat it as a Markdown file.

Being able to control what file extensions should be processed in what way by Astro would not only solve aforementioned issues, but also make Astro more future proof. It could potentially serve as a basis for allowing Astro community to build custom file processors in the future, which they currently can only do by [hacking the private API in setup hook](https://github.com/shishkin/astro-asciidoc).

# Goals

- Allow assignment of a processor (Markdown, MDX, MarkDoc, AsciiDoc, etc.) to a file extension (`.md`, `.mdx`, `.mdoc`, `.adoc`, `.lhs`, `.foo`, etc.)

# Example
I'm not sure what would be the ideal way to do this, but it could be done by allowing user to pass an object mapping somewhere rougly like this.

```typescript
{
  md: MarkdownProcessor,
  mdox: MarkdocProcessor,
  adoc: AsciiDocProcessor,
  lhs: MarkdownProcessor,
  foo: FooBarProcessor,
}
```

**Comment by airtonix** [comment]
Maybe i don't know enough about Astro, but doesn't it just defer to vite for this? 

So this should possible via vite-plugins that transform files? 

https://vitejs.dev/guide/api-plugin.html#transforming-custom-file-types

**Reply by kamoshi** [reply]
As far as I can tell, this is used for a different thing, i.e. transforming files imported directly in the source code of a Vite project, so it wouldn't help in this case.

The situation it would be useful in is if I wanted to, for example, use ReScript language together with Astro instead of TypeScript.

Here, the problem is that Astro is hardcoded to only look for content files like `.md`, `.mdx` or `.mdoc`, it's a different kind of an issue.

**Comment by bluwy** [comment]
I'm not really sure if Astro should allow this. Each extension should mean a feature by itself. For example, if you want to write MarkDoc, you should use `.mdoc`, not disguising as `.md`. This would also affect syntax highlighting (and possibly other tools) by default.

If you want to use Asciidoc for example, then you could create an integration like `@astrojs/mdx` so that `.mdx` files are "understood" and processed by Astro.

**Reply by kamoshi** [reply]
> Each extension should mean a feature by itself.

That's how it would work ideally, but that's not how it works in reality. Please, consider the fact that not even an `.md` file is universal, because some programs consider `.md` to be [CommonMark](https://commonmark.org/), while some others consider it to be [Github Flavored Markdown](https://github.github.com/gfm/), while yet again some others include i.e. wikilinks as a nonstandard extension, such as in Obsidian and other Zettelkasten-like software.

>  if you want to write MarkDoc, you should use .mdoc, not disguising as .md

I personally feel like this decision is something, which would be better left for a user to make, because otherwise it causes a whole lot of issues down the line, which I mentioned in the OP.

> This would also affect syntax highlighting (and possibly other tools) by default.

You can change how each editor should interpret each file extension, for instance you can force VS Code or Neovim to treat `.mdx` as a markdown file, which they don't do out of the box.

> If you want to use Asciidoc for example, then you could create an integration like @astrojs/mdx so that .mdx files are "understood" and processed by Astro.

The problem is that this is currently *not* possible, because the functionality is *not* exposed in public API, so every attempt to do this is a *quite big* hack, which could break at any time in any future update. There is neither a way to set an extension to be interpreted in an X way by Astro, nor is there a way to write an integration for Astro which would add a way to process additional file extension in an Y way.

Currently the way Astro works is that there is hardcoded functionality to process:
- `.md` - markdown, but with Remark, so it alleviates some issues, as Remark lets you redefine what Markdown is
- `.mdx` - MDX 
- `.mdoc` - Markdoc

If you find yourself having some files which do not fit these three hardcoded "slots", then you're simply in trouble, and have to hack around it by for example using hard links .i.e `myFile.lhs` -> `myFile.md` just to force Astro to read this file, which is pretty annoying to say the least.

**Reply by bluwy** [reply]
> The problem is that this is currently _not_ possible, because the functionality is _not_ exposed in public API, so every attempt to do this is a _quite big_ hack, which could break at any time in any future update. There is neither a way to set an extension to be interpreted in an X way by Astro, nor is there a way to write an integration for Astro which would add a way to process additional file extension in an Y way.

If this is not possible, we should definitely fix it. But we do expose the right APIs in order to support more "content"-style extensions. That's why we have `@astrojs/mdx` and `@astrojs/markdoc` separated out as packages. You can flag it as an issue if you see MDX or Markdoc hardcoded in Astro core, and we can try to fix it.

**Reply by kamoshi** [reply]
I've seen the source code of those extensions, when I was trying to use that API for my own stuff 😄

The setup hook object has two [hidden properties](https://github.com/withastro/astro/blob/e4999083bf3d0f41c8e2a4f44d19b02d95529ec6/packages/integrations/mdx/src/index.ts#L31):
```ts
type SetupHookParams = HookParameters<'astro:config:setup'> & {
	// `addPageExtension` and `contentEntryType` are not a public APIs
	// Add type defs here
	addPageExtension: (extension: string) => void;
	addContentEntryType: (contentEntryType: ContentEntryType) => void;
};
```

This is how it is [used](https://github.com/withastro/astro/blob/e4999083bf3d0f41c8e2a4f44d19b02d95529ec6/packages/integrations/mdx/src/index.ts#L43-L58):
```ts
const { updateConfig, config, addPageExtension, addContentEntryType, addRenderer } =
	params as SetupHookParams;

addRenderer(astroJSXRenderer);
addPageExtension('.mdx');
addContentEntryType({
	extensions: ['.mdx'],
	async getEntryInfo({ fileUrl, contents }: { fileUrl: URL; contents: string }) {
		const parsed = parseFrontmatter(contents, fileURLToPath(fileUrl));
		return {
			data: parsed.data,
			body: parsed.content,
			slug: parsed.data.slug,
			rawData: parsed.matter,
		};
	},
```

This effectively means that while it is *technically* possible to add more usable file extensions, none of those functions can *realistically* be used (trusted) to add more file extensions, other than the officially supported Markdoc and MDX integrations, because the contract is that any and all changes to Astro might break this at any time, and on top of that, there's is no documentation for this outside the source code itself
